File: Particles_in_BH_Tree_III/ngb_project/initial_assess.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <iterator>

int main() {
    const std::string filename = "h_child.bin";
    
    // Open the file in binary mode
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        std::cerr << "Error opening file." << std::endl;
        return 1;
    }

    // Determine the size of the file
    file.seekg(0, std::ios::end);
    std::streamsize size = file.tellg();
    file.seekg(0, std::ios::beg);

    // Calculate the number of integers
    int num_ints = size / sizeof(int);

    // Read the file into a vector
    std::vector<int> h_child(num_ints);
    if (file.read(reinterpret_cast<char*>(h_child.data()), size)) {
        std::cout << "h_child.shape = " << h_child.size() << std::endl;
    } else {
        std::cerr << "Error reading file." << std::endl;
        return 1;
    }

    // Find the last index where h_child is not -1
    int lastIndex = -1;  // Initialize to -1 to indicate not found
    for (int i = 0; i < num_ints; ++i) {
        if (h_child[i] != -1) {
            lastIndex = i;
        }
    }

    std::cout << "Last index where h_child is not -1: " << lastIndex << std::endl;
    
    if (lastIndex > static_cast<int>(num_ints - 0.1f * num_ints))
      std::cout << "Danger !!!! It seems the h_child (and d_child) size is not enough. Try to increase it !!!" << std::endl;



    return 0;
}

File: Particles_in_BH_Tree_III/ngb_project/initial_assess.py
import struct
import numpy as np

filename = "h_child.bin"

with open(filename, 'rb') as file:
    # Read the entire file into a bytes object
    data = file.read()

# Calculate the number of integers in the file
num_ints = len(data) // 4  # 4 bytes per int

# Unpack the data into integers
h_child = np.array(struct.unpack(f'{num_ints}i', data))

print("h_child.shape = ", h_child.shape)

nx = np.where(h_child != -1)[0]

print("len(nx) = ", len(nx))
print()
print("nx = ", nx)

